Rewrite tutorials in Markdown

mrgiles opened the pull request:
Rewrite [kubernetes-basics](https://kubernetes.io/docs/tutorials/kubernetes-basics/) tutorials (HTML) in Markdown

Fixes: #41625

mrgiles commented:
Please let me know if I'm on the right track here. 
I've found some HTML formatting that I couldn't replicate in Markdown.
Thanks!

mrgiles commented:
List of tutorials to rewrite:

```
kubernetes-basics
├── create-cluster
├── deploy-app
├── explore
├── expose
├── scale
└── update
```

dipesh-rawat commented on content/en/docs/tutorials/kubernetes-basics/_index.md:
```diff
@@ -0,0 +1,56 @@
+is a production-ready, open source platform designed with Google's accumulated experience in
+container orchestration, combined with best-of-breed ideas from the community.
+
+## Kubernetes Basic Modules
```
I think this current sub-section [(preview here)](https://deploy-preview-48592--kubernetes-io-main-staging.netlify.app/docs/tutorials/kubernetes-basics/#kubernetes-basic-modules) of the list of tutorial pages feels a bit confusing with the links and images. It might be more effective to use a simple bullet list of links without images. Just my first thought looking at the [page](https://deploy-preview-48592--kubernetes-io-main-staging.netlify.app/docs/tutorials/kubernetes-basics/) - I’m sure others may have different opinions on this.

mrgiles replied:
Yes, I don't like it, either but I didn't know how much leeway I have to change the original content.

mrgiles replied:
I'll upload a draft with just text links

mrgiles commented on content/en/docs/tutorials/kubernetes-basics/_index.md:
```diff
@@ -0,0 +1,56 @@
+no_list: true
+weight: 20
+content_type: concept
+---
```
Got it. Thanks for reviewing, @dipesh-rawat.

mrgiles commented:
I'm not sure how to replicate/replace those blue boxes with definitions on the right side of the pages

mrgiles commented:
I'm going to stop at 3 of 6 tutorials, review/edit them with your feedback, and then add the other 3. They're all similar in length. Thanks.

mrgiles commented:
Edited, so far:

```
kubernetes-basics
├── create-cluster
├── deploy-app
└── explore
```

mrgiles commented:
I'm adding the 4th tutorial in the series. 

Edited, so far: 

```
kubernetes-basics
├── create-cluster
├── deploy-app
├── explore
└── expose
```

mrgiles commented:
Adding the 5th tutorial in the series.

Edited, so far:

```
kubernetes-basics
├── create-cluster
├── deploy-app
├── explore
├── expose
└── scale
```

mrgiles commented:
Added the 6th, and last, tutorial in the **kubernetes-basics** series:

```
kubernetes-basics
├── create-cluster
├── deploy-app
├── explore
├── expose
├── scale
└── update
```
I will now focus on reviewing the documents, fixing typos, styles, etc. 
I still have to figure out how to incorporate those blue text boxes on the right margins. 
Use Note for those? Suggestions?
Thanks!

mrgiles commented:
I hope you like it @stormqueen1990, this is to fix your Issue! :)

mrgiles commented:
@sftim, any suggestions on what to do with those blue text boxes on the right margins? I don't know how to replicate them in Markdown. I thought of just using **Note** for them.

Also, for the image animations (`carousel`), I just placed the individual images in numerical sequence on the page, instead of trying to replicate the animations, which seem to require additional CSS and/or JS to embed them in Markdown.

stormqueen1990 reviewed (commented):
Hi there, @mrgiles! I'm not a reviewer for English so this is just a piece of feedback I found important to write down!

mrgiles commented on content/en/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro.md:
```diff
@@ -0,0 +1,64 @@
+---
+title: Using Minikube to Create a Cluster
+weight: 10
```
Done. Thanks for checking!

mrgiles commented:
Hi @stormqueen1990, I see a problem by adding these descriptions, it also added them under the title of each page and they become a duplicate of the Objectives section content. Now I have the same info in two consecutive sections. Please see the preview. I think that's why I deleted those descriptions initially. I feel like I should delete them again. What do you think? Thanks!

Andygol commented on content/en/docs/tutorials/kubernetes-basics/_index.md:
```diff
@@ -0,0 +1,60 @@
+
+## Kubernetes Basic Modules
+
+### [1. Create a Kubernetes cluster](/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro/)
+
+<!-- {{< figure src="public/images/module_01.svg" class="diagram-small" >}} -->
+
+### [2. Deploy an app](/docs/tutorials/kubernetes-basics/deploy-app/deploy-intro/)
+
+<!-- {{< figure src="public/images/module_02.svg" class="diagram-small" >}} -->
+
+### [3. Explore your app](/docs/tutorials/kubernetes-basics/explore/explore-intro/)
+
+<!-- {{< figure src="public/images/module_03.svg" class="diagram-small" >}} -->
+
+### [4. Expose your app publicly](/docs/tutorials/kubernetes-basics/expose/expose-intro/)
+
+<!-- {{< figure src="public/images/module_04.svg" class="diagram-small" >}} -->
+
+### [5. Scale up your app](/docs/tutorials/kubernetes-basics/scale/scale-intro/)
+
+<!-- {{< figure src="public/images/module_05.svg" class="diagram-small" >}} -->
+
+### [6. Update your app](/docs/tutorials/kubernetes-basics/update/update-intro/)
+
+<!-- {{< figure src="public/images/module_06.svg" class="diagram-small" >}} -->
```
It also possible to replicate such layout using table

<img width="573" alt="image" src="https://github.com/user-attachments/assets/6df31166-ef5a-4235-9204-d695afd80ed5">

```suggestion
{{< figure src="public/images/module_01.svg" class="diagram-small" >}} | {{< figure src="public/images/module_02.svg" class="diagram-small" >}} | {{< figure src="public/images/module_03.svg" class="diagram-small" >}}
--- | --- | ---
**[1. Create a Kubernetes cluster](/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro/)** | **[2. Deploy an app](/docs/tutorials/kubernetes-basics/deploy-app/deploy-intro/)** | **[3. Explore your app](/docs/tutorials/kubernetes-basics/explore/explore-intro/)**
{{< figure src="public/images/module_04.svg" class="diagram-small" >}} | {{< figure src="public/images/module_05.svg" class="diagram-small" >}} | {{< figure src="public/images/module_06.svg" class="diagram-small" >}}
**[4. Expose your app publicly](/docs/tutorials/kubernetes-basics/expose/expose-intro/)** | **[5. Scale up your app](/docs/tutorials/kubernetes-basics/scale/scale-intro/)** | **[6. Update your app](/docs/tutorials/kubernetes-basics/update/update-intro/)**
```

mrgiles replied:
Hi @Andygol. Thank you for reviewing.
I think I'll go with the simple numbered list suggestion.
I tried the table approach but the default row background color alternates white/gray, which doesn't look great.
To change this requires some overly complicated CSS code. Plus, the rendered images sizes look a little odd. 

![2024-11-13-01](https://github.com/user-attachments/assets/72cf6528-8df2-4e79-96a3-e5448843d780)

mrgiles replied:
Solved using [CSS Flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_flexible_box_layout/Basic_concepts_of_flexbox). Hope this is allowed here?

mrgiles commented:
@sftim, any suggestions on what to do with these blue text boxes on the right margins? I don't know how to replicate them in Markdown. I thought of replacing them with `{{< note >}}` entries. What do you think?

![2024-11-14-131912_hyprshot](https://github.com/user-attachments/assets/104afcee-19da-403e-9699-94333b63b613)

Also, for the [current](https://kubernetes.io/docs/tutorials/kubernetes-basics/scale/scale-intro/) image animations (carousel), I just placed the individual images [proposed](https://deploy-preview-48592--kubernetes-io-main-staging.netlify.app/docs/tutorials/kubernetes-basics/update/update-intro/) in numerical sequence on the page, instead of trying to replicate the animations, which seem to require additional CSS and/or JS to embed them in Markdown.

Does that work for you or should we try to recreate the animations with CSS+JS? (Which I don't really know how to do it.)

sftim commented:
I would take the question to #sig-docs. Try to pitch this for people who enjoy a web development challenge?

mrgiles commented:
I'm testing two options for the blue boxes (shown in the `cluster-intro.md `page here). To make them look like in the original docs, would require to rewrite most of the pages again in HTML+CSS, which would defeat the primary objective of issue #48592, that was to rewrite them in Markdown. This issue looks like a catch-22 situation.

mrgiles commented:
Okay, I managed to restore the animations (turns out it was just a matter of reusing the existing CSS code).

Now I need somebody to confirm which one of the two proposed "blue boxes" options you'd like to use.
Like I've explained above, trying to replicate the original design of having the blue boxes on the right margin would require to practically rewrite everything in HTML, which would defeat the purpose of the original issue that was to rewrite it in Markdown.

But I guess it depends on how important it is to preserve 100% of the original layout. 
I'll ask in Slack to have somebody chime in. Thanks.

sftim commented:
> preserve 100% of the original layout

IMO: not important

sftim commented on content/en/docs/tutorials/kubernetes-basics/_index.md:
```diff
@@ -0,0 +1,81 @@
+## Kubernetes Basics Modules
+
+<!-- this css code is to preserve the original format -->
+<div style="display: flex; flex-wrap: wrap; gap: 20px; justify-content: space-between;">
+  <div style="width: 30%; text-align: center;">
```
Ideally, use CSS classes rather than inline CSS.

mrgiles replied:
I moved the classes to a separate css file.

sftim commented on content/en/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro.md:
```diff
@@ -0,0 +1,78 @@
+
+## Kubernetes Clusters
+
+<div class="col-md-8">
+  <div class="content__box content__box_fill">
+    <p><i>Kubernetes is a production-grade, open-source platform that orchestrates
+      the placement (scheduling) and execution of application containers
+      within and across computer clusters.</i></p>
+  </div>
+</div>
```
I would either use this:

```suggestion
{{% alert title="What is Kubernetes" color="primary" %}}
      Kubernetes is a production-grade, open-source platform that orchestrates
      the placement (scheduling) and execution of application containers
      within and across computer clusters.
{{% /alert %}}
```
_or_ make a similar but more custom shortcode

_or_ a plain paragraph with a custom class (and some associated CSS).

mrgiles replied:
I've used the `{{% alert %}}` suggestion.

sftim commented on content/en/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro.md:
```diff
@@ -0,0 +1,78 @@
+coordinates all activities in your cluster, such as scheduling applications, maintaining
+applications' desired state, scaling applications, and rolling out new updates.
+
+<div class="col-md-8">
+  <div style="background-color: #eaf7ff; border-left: 5px solid #3771e3; padding: 15px; margin: 20px 0;">
+    <p><i>Control Planes manage the cluster and the nodes that are used to host the running applications.</i></p>
+  </div>
+</div>
```
Is it feasible to use a Markdown paragraph with a custom class (and some associated CSS)?

mrgiles replied:
I've used the `{{% alert %}}` suggestion.

sftim commented on content/en/docs/tutorials/kubernetes-basics/explore/explore-intro.md:
```diff
@@ -0,0 +1,188 @@
+
+### Nodes overview
+
+{{< figure src="../../public/images/module_03_nodes.svg" class="diagram-medium" >}}
```
Use site relative links (avoid starting with `../`)

mrgiles replied:
Done

sftim commented on content/en/docs/tutorials/kubernetes-basics/create-cluster/cluster-intro.md:
```diff
@@ -0,0 +1,78 @@
+
+### Cluster Diagram
+
+{{< figure src="../../public/images/module_01_cluster.svg" class="diagram-medium" >}}
```
Use site relative links (avoid starting with `../`)

mrgiles replied:
done

sftim reviewed (commented):
This looks OK. It'd be great to drop even more of the HTML, though.

mrgiles commented:
Call for reviewers: @reylejano, @dipesh-rawat, @tengqm. If/when any of you have some time to review these changes?

I know that everybody has been very busy with KubeCon and the recent K8s release, so maybe now you have some spare cycles?

I've tried to rewrite the tutorials in a way that resemble the original HTML docs but some formatting doesn't have a 1:1 equivalent in Markdown so I had to use a little HTML+CSS to approximate the original look.

Thanks!

tengqm commented:
overall lgtm.
I don't like the embedding of CSS in markdown, but that can be left to a future PR.
/lgtm

mrgiles commented:
/assign @sftim

sftim commented on static/css/style_tutorials.css:
I disagree with adding this.

We have a wider piece of work aimed at moving _all_ our stylesheets into `assets/scss` (this also allows minifying on build, and as a side effect we verify that the built CSS is syntactically valid).

@mrgiles if you agree with the suggestion I'm about to make: I can push a commit to my fork that shows an alternative way to achieve the same outcome, without putting any CSS in as static content.

How does that sound?

mrgiles replied:
Sure @sftim. Sorry, I wasn't aware of that wider effort that you mention. You can also point me to some relevant code snippets that I can adapt to this solve this particular case. Whatever is easier for you, I know you're very busy. Thanks!

sftim replied:
https://gohugo.io/content-management/markdown-attributes/ explains how to set a CSS class on a paragraph, which I think can help.
You can also use [shortcodes](https://gohugo.io/content-management/shortcodes/) that process their inner content.

For an example of the kind of change, have a look at https://github.com/kubernetes/website/pull/48363/files

specifically
- the changes to `assets/scss/_styles_project.scss`
- the changes to `layouts/case-studies/baseof.html`

sftim commented:
@mrgiles as there's been a few changes since you opened this PR, could you:

- rebase against _main_
- squash to only a few commits

?

tengqm commented:
This is a huge effort. I'd like to vote for kicking it in and then iterate.

/approve
